presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometrySerde.java

```diff
@@ -150,51 +148,94 @@ private static OGCGeometry readSimpleGeometry(BasicSliceInput input, Slice input
-                        envelope = new Envelope(xMin, yMin, xMax, yMax);
-                    }
-                }
-                if (envelope != null) {
-                    if (overallEnvelope == null) {
-                        overallEnvelope = envelope;
-                    }
-                    else {
-                        overallEnvelope.merge(envelope);
-                    }
-                }
-            }
+        int length = input.available() - 1;
+        GeometryType type = GeometryType.getForCode(input.readByte());
+        return getEnvelope(input, type, length);
+    }
+
+    private static Envelope getEnvelope(BasicSliceInput input, GeometryType type, int length)
+    {
+        switch (type) {
+            case POINT:
+                return getPointEnvelope(input);
+            case MULTI_POINT:
+            case LINE_STRING:
+            case MULTI_LINE_STRING:
+            case POLYGON:
+            case MULTI_POLYGON:
+                return getSimpleGeometryEnvelope(input, length);
+            case GEOMETRY_COLLECTION:
+                return getGeometryCollectionOverallEnvelope(input);
+            default:
+                throw new UnsupportedOperationException("Unexpected type: " + type);
         }
+    }
 
+    private static Envelope getGeometryCollectionOverallEnvelope(BasicSliceInput input)
+    {
+        Envelope overallEnvelope = null;
+        while (input.available() > 0) {
+            int length = input.readInt() - 1;
+            GeometryType type = GeometryType.getForCode(input.readByte());
+            Envelope envelope = getEnvelope(input, type, length);
+            overallEnvelope = merge(overallEnvelope, envelope);
+        }
         return overallEnvelope;
     }
+
+    private static Envelope getSimpleGeometryEnvelope(BasicSliceInput input, int length)
+    {
+        // skip type injected by esri
+        input.readInt();
+
+        double xMin = input.readDouble();
+        double yMin = input.readDouble();
+        double xMax = input.readDouble();
+        double yMax = input.readDouble();
+
+        int skipLength = length - (4 * Double.BYTES) - Integer.BYTES;
+        verify(input.skip(skipLength) == skipLength);
+
+        if (isEsriNaN(xMin) || isEsriNaN(yMin) || isEsriNaN(xMax) || isEsriNaN(yMax)) {
+            // TODO: isn't it better to return empty envelope instead?
+            return null;
+        }
+        return new Envelope(xMin, yMin, xMax, yMax);
+    }
+
+    private static Envelope getPointEnvelope(BasicSliceInput input)
+    {
+        // skip type injected by esri
+        input.readInt();
+
+        double x = input.readDouble();
+        double y = input.readDouble();
+        if (isEsriNaN(x) || isEsriNaN(y)) {
+            // TODO: isn't it better to return empty envelope instead?
+            return null;
+        }
+        return new Envelope(x, y, x, y);
+    }
+
+    @Nullable
+    private static Envelope merge(Envelope left, @Nullable Envelope right)
```

Comment:
`left` is not marked as `@Nullable`, but it can be null as we check whether it's null below. Also I am not sure whether we use `Nullable` for method parameters, so I guess you can just omit them here (IIRC we use `Nonnull` if they cannot be null).

Reply:
We use this annotation to annotate parameters (e.g. `SessionPropertyManager#decodeCatalogProperties`). I though we assume by default that parameters are not null. If there's a parameter that could be null - we use `Optional<?>`. But maybe i'm wrong. @electrum @dain  ?